src/core/Track.cpp

```diff
@@ -561,13 +564,15 @@ bar_t Track::length() const
 		}
 
 		const tick_t cur = clip->endPosition();
-		if( cur > last )
+		if (cur > last)
 		{
 			last = cur;
 		}
 	}
 
-	return last / TimePos::ticksPerBar();
+	m_length = last / TimePos::ticksPerBar();
+	
+	if (Engine::getSong()) { Engine::getSong()->updateLength(); };
```

Comment:
I don't think we need the `if` here? We should always have a Song loaded I believe.

Reply:
I think you're right, I just got a bit nervous because `Track::removeClip` checks for a song for some reason.

Reply:
Yeah I'm not sure why. Maybe if doing a command line render? But even then we should have a song. Maybe @PhysSong has an idea?

Reply:
> `Track::removeClip` checks for a song for some reason

The check was introduced in 2010 by commit 488337f6eb7d4c52d2d143d7fa23b5a3cca92f9f, but I have no ideas why this has been added.

Reply:
Alright, good to know that it's unknown. 😄

Reply:
Found it.  LMMS segfaults when you close the program if you don't check for the Song here.  Turns out it is possible for these functions to run without an active Song, when LMMS is closing itself.

Reply:
> The check was introduced in 2010 by commit [488337f](https://github.com/LMMS/lmms/commit/488337f6eb7d4c52d2d143d7fa23b5a3cca92f9f), but I have no ideas why this has been added.

I now understand why it's required. The `deleteHelper` function sets `Engine::s_song` to `nullptr` before deleting the instance, so the check is required.
The intention of `deleteHelper` is to detect use-after-free, but it also prevents getting the instance being deleted via `Engine`, which may lead to crashes.